core/modules/qos.cc

```diff
@@ -0,0 +1,441 @@
+
+      size_t len = reinterpret_cast<size_t>(total_key_size_ / sizeof(uint64_t));
+
+      for (size_t i = 0; i < len; i++) {
+        keys[j].u64_arr[i] = keys[j].u64_arr[i] & mask[i];
+      }
+    }
+  }
+
+  uint64_t hit_mask = table_.Find(keys, val, cnt);
+
+  for (int init = 0; init < cnt; init++) {
+    if ((hit_mask & (1ULL << init)) == 0) {
+      EmitPacket(ctx, batch->pkts()[init], default_gate);
+      // std::cout << "qer default gate " << default_gate << std::endl;
+      continue;
+    }
+
+    /* if lookup was successful, then set values (if possible) */
+    if (hit_mask & (1 << init)) {
+      // std::cout << "qer gate " << val[init]->ogate << std::endl;
+      if ((val[init]->ogate == CONFIG_DROP_GATE) ||
+          (val[init]->ogate == METER_GREEN_GATE) ||
+          (val[init]->ogate == METER_YELLOW_GATE) ||
+          (val[init]->ogate == METER_RED_GATE)) {
+        // std::cout << "qer drop " << val[init]->ogate << std::endl;
+        EmitPacket(ctx, batch->pkts()[init], CONFIG_DROP_GATE);
+      } else if (val[init]->ogate == UNMETERED_GATE) {
+        // std::cout << "qer unmeter " << val[init]->ogate << std::endl;
+        EmitPacket(ctx, batch->pkts()[init], UNMETERED_GATE);
+      } else if (val[init]->ogate == METER_GATE) {
+        // std::cout << "qer meter before" << val[init]->ogate << std::endl;
+        uint64_t time = rte_rdtsc();
+        uint16_t ogate = CONFIG_DROP_GATE;
+        uint8_t color = rte_meter_trtcm_color_blind_check(
+            &val[init]->m, &val[init]->p, time,
+            batch->pkts()[init]->total_len());
+
+        // std::cout << "color : " << color << std::endl;
+        pkt = batch->pkts()[0 + init];
+        size_t num_values_ = values_.size();
+        if (color == RTE_COLOR_GREEN) {
+          ogate = METER_GREEN_GATE;
+        } else if (color == RTE_COLOR_YELLOW) {
+          ogate = METER_YELLOW_GATE;
+        } else if (color == RTE_COLOR_RED) {
+          ogate = METER_RED_GATE;
+        }
+        // std::cout << "qer meter after" << val[init]->ogate << std::endl;
+        for (size_t i = 0; i < num_values_; i++) {
+          int value_size = values_[i].size;
+          int value_pos = values_[i].pos;
+          int value_off = values_[i].offset;
+          int value_attr_id = values_[i].attr_id;
+          uint8_t *data = pkt->head_data<uint8_t *>() + value_off;
+
+          if (value_attr_id < 0) { /* if it is offset-based */
+            memcpy(data,
+                   reinterpret_cast<uint8_t *>(&(val[init]->Data)) + value_pos,
+                   value_size);
+          } else { /* if it is attribute-based */
+            typedef struct {
+              uint8_t bytes[bess::metadata::kMetadataAttrMaxSize];
+            } value_t;
+            uint8_t *buf = (uint8_t *)(&(val[init]->Data)) + value_pos;
+
+            DLOG(INFO) << "Setting value " << std::hex
+                       << *(reinterpret_cast<uint64_t *>(buf))
```

Comment:
this dlog info does'nt interpret 'buf' correctly , if the value_size is other then uint64_t. correct buf can be printed from switch-case ,below as well.